[1단계 - 출석 구현] 히스타(오승현) 미션 제출합니다.

Hacanna42 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

피케이 안녕하세요! 리뷰 해주셔서 감사합니다 :)
처음으로 TDD를 학습하고 적용해보며 미션을 진행해봤어요.
첫 시작때는 TDD를 정말 작은 단위로 철저하게 지키면서 진행했는데요, 사실 진행하다보니 이대로는 절대 마감 시간 안에 구현을 못 마칠 것 같아서 TDD를 어느정도 포기하게 되었어요.

그래도 TDD가 첫 갈피를 잡아준 것 같아서 어느정도 장점은 알겠는데, 이렇게 마감 기한이 짧은 상황에서도 적용할 수 있을지는 의문이에요..!

그리고 이번 미션에서 답이 안나오는 고민되는 사항은 없었는데.. 설계가 너무 어려워서 힘들었어요.. 이 설계가 맞나..? 라는 생각이 계속 들었어요. 그게 제일 힘들었던 것 같아요.

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->
  
 프롤로그 링크: https://prolog.techcourse.co.kr/studylogs/3931


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

<!-- 이유 작성 -->
시간 부족으로 인해 리팩터링이 부족했고, depth가 1 초과인 부분이 있었어요.
또, 설계 부분에서 좋은 객체 지향 설계가 맞는지 의문이 들었어요.

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->
객체지향 설계에 집중하여 리뷰해주시면 감사하겠습니다 :)

Hacanna42 committed:
test, feat: 지각 여부 판단 TDD 사이클 완료

Hacanna42 committed:
test, feat: 결석 여부 판단 TDD 사이클 완료

Hacanna42 committed:
test, feat: 휴일 체크 TDD 사이클 완료

Hacanna42 committed:
test, feat: 일자에 따라 요일 반환 기능 TDD 사이클 완료

Hacanna42 committed:
test, feat: 클래스 분리 및 테스트 코드 이동

Hacanna42 committed:
test, feat: 출석 타입 판단 TDD 사이클 완료

Hacanna42 committed:
docs(readme): 요구사항 정리

Hacanna42 committed:
docs(readme): 요구사항 이쁘게

Hacanna42 committed:
test, feat: Function 열거형 TDD 사이클 완료

Hacanna42 committed:
test, feat: Crew 구현 TDD 사이클 완료

Hacanna42 committed:
test, feat: Parser 시각 파싱 구현 TDD 사이클 완료

Hacanna42 committed:
test, feat: 날짜 관련 구현 TDD 사이클 완료

Hacanna42 committed:
feat: 출석 확인 기능 구현

Hacanna42 committed:
feat: 날짜 및 시간 입력 구현

Hacanna42 committed:
docs: 기능 구현 진행 상황 업데이트

Hacanna42 committed:
feat: 출석 기능 구현

Hacanna42 committed:
test, feat: 날짜 예외 처리

Hacanna42 committed:
feat: 출석 확인 출력 기능 구현

Hacanna42 committed:
feat: 출석 수정 출력 구현

Hacanna42 committed:
feat: 종료 기능 구현

Hacanna42 committed:
feat: 크루별 출석 기록 확인 구현

Hacanna42 committed:
feat: 제적 위험자 조회 구현

Hacanna42 committed:
feat: .csv 파일 읽기 구현

Hacanna42 committed:
feat: 제적 위험자 정렬 구현

Hacanna42 committed:
refactor: 불필요한 임포트 구문, 개행 삭제

Hacanna42 committed:
refactor: 불필요한 필드 및 함수 삭제

Hacanna42 committed:
refactor: 패키지 구조 변경

pkeugine commented on README.md:
```diff
@@ -1,3 +1,63 @@
+
+### 📅 출석 일자
+- [x] 출석 일자를 저장하는 역할
+    - 요일을 반환하는 메서드
```
이 부분이 구현되어 있지 않습니다!
> 오늘은 12월 14일 토요일입니다. 기능을 선택해 주세요.

미션의 요구사항을 보면 위 내용이 기능 선택 전에 위치해야 하며 요일에 따라 아래 에러를 발생시켜야 합니다.
> [ERROR] 12월 14일 토요일은 등교일이 아닙니다.

요구사항에 맞는 기능을 구현하는 것이 가장 중요한 일입니다. TDD가 이를 놓치지 않게 하는 장치가 되고요. 
특히 예외가 발생하는 케이스에 대한 기능 확인과 단위 테스트 추가는 빠트리지 않게 주의해주세요! 😀

Hacanna42 replied:
네. 미션 설명에서는 12월 시범 사업인데, 개발하는 현재 날짜가 2월이라서 처음에 날짜를 어떻게 받아올지 고민했었어요. 그래서 결국 매 기능마다 적용할 날짜를 입력받는것으로 타협을 보았는데, 이제는 프로그램 첫 시작 때 현재 날짜를 수동으로 입력받도록 수정했어요. 따라서  

> 오늘은 12월 14일 토요일입니다. 기능을 선택해 주세요.

부분을 구현할 수 있게 되었고, 구현 완료했습니다.

> [ERROR] 12월 14일 토요일은 등교일이 아닙니다.

이 부분은 제가 확실히 놓친 부분이에요. 시간에 쫒기다 보니 너무 많은 걸 놓쳤네요... 확실하게 짚어주셔서 감사합니다 피케이!!

pkeugine commented on src/main/java/controller/dto/AttendanceHistoryDto.java:
```diff
@@ -0,0 +1,24 @@
+) {
+    public static AttendanceHistoryDto from(AttendanceDateTime attendanceDateTime) {
+        return new AttendanceHistoryDto(
+                attendanceDateTime.getMonth(),
```
현재 `attendanceDateTime.getMonth()`를 호출하고 이 메서드는 다시 [attendanceDate.getMonth()](url)를 호출하고 있는데, `attendanceDate`의 `month` 값이 결국 상수 12로 고정되어 있습니다.🤔
불필요한 호출 없이 더 효율적인 방식으로 개선할 수 있을지 고민해보면 좋을 것 같아요!

Hacanna42 replied:
Date와 Time을 DateTime이라는 새로운 클래스에 래핑해놓아서, 구조가 복잡해진 부분이 있는 것 같아요..!
그렇다고 Date, Time 두개로 나누지 않고 DateTime에서 날짜와 시간에 관련한 로직을 다 처리하면 역할이 너무 많아지는 것 같기도 하고요.. 많이 고민되는 부분인 것 같아요..! 

혹시 개선할 수 있는 방법이 있다면 아이디어를 주실 수 있으신가요?

pkeugine replied:
전에 코멘트로 말씀드렸던 것처럼 저라면 `java.time.LocalDate`, `java.time.LocalDateTime` 등 자바에서 제공해주는 클래스를 활용해볼 것 같네요!

pkeugine commented on src/main/java/domain/attendance/AttendanceHistory.java:
```diff
@@ -0,0 +1,42 @@
+    private final Crew crew;
+    private final AttendanceDateTime attendanceDateTime;
+
+    private AttendanceHistory(Crew crew, AttendanceDateTime attendanceDateTime, AttendanceType attendanceType) {
```
사용하지 않는 `attendanceType`를 받네요!
파라미터 `attendanceType`를 제거하면 생성자와 동일한 파라미터를 받는 `of`가 불필요해질 것 같아요

Hacanna42 replied:
사용하지 않는 파라미터를 받는 부분은 미처 리팩토링 하지 못한 부분이네요. 짚어주셔서 감사합니다!
`attendanceType` 을 제거하면, 정적 팩토리 메서드 `of`가 불필요해질 것 같다고 말씀해주셨는데, 
저는 아래 근거들로 정적 팩토리 메서드를 사용하는 걸 조금 더 선호하는 편이에요.

1. 생성 관련 코드를 관련 클래스가 갖게 되면서 코드가 깔끔해진다.
2. 생성자가 이름을 가지게 되면서 네이밍이 명확해진다.
3. 외부에서 해당 클래스의 객체를 생성할 때, 생성자에 어떤 파라미터들이 필요한 지 종류별로 즉각적으로 확인할 수 있다.

그래서 저는 생성자와 동일한 파라미터를 가진다고 해도 정적 팩토리 메서드를 삭제하고 싶진 않는데, 정적 팩토리 메서드가 불필요해질 것 같다는 건 삭제를 권장한다는 말씀이신가요? 만약 그렇다면, 이유가 무엇인가요?

pkeugine replied:
히스타가 말해주신 내용들 좋습니다 👍
그런데 ‘네이밍이 명확해진다‘고 하려면 그 이름이 ’of' 보다는 어떤 로직을 통해 생성되는지 담으면 좋을 듯 하네요.

생성자와 정적팩토리가 혼용되어 사용되고, 그 수가 많아지는 경우를 생각해보죠. 변경점이 생겼을 때 수정할 부분이 그 만큼 더 많아집니다. 그렇기 때문에 저는 객체 생성 방식을 다양하게 열어두는 것을 지양해요.

삭제를 권장한다고 그게 항상 정답은 아닙니다. 히스타 스스로, 그리고 함께할 팀원이 그 방식이 맞다고 생각이 든다면 유지하셔도 괜찮습니다.

pkeugine commented on src/main/java/domain/attendance/AttendanceHistory.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    public boolean isSameCrew(AttendanceHistory attendanceHistory) {
+        return crew.equals(attendanceHistory.crew);
```
`this` 키워드를 모두 작성해줄지, 불필요할 때에는 제거할지를 통일하는 것이 좋습니다!

Hacanna42 replied:
좋은 피드백 감사드립니다! 혹시 제가 이해한 게 맞는지 확인해주실 수 있으실까요?
현재 `AttendanceHistory`의 모든 코드에는 this 키워드가 꼭 필요한 경우에만 들어가 있는데요.

>  `this`키워드를 모두 작성해줄지, 불필요할 때에는 제거할지를 통일하는 것이 좋습니다!

라는 말씀은, `crew`앞에 `this`가 불필요함에도 불구하고 읽다 보면 헷갈릴 수 있기 때문에, 이렇게 `this` 키워드를 썼다 안썼다 하는 상황에서는 차라리 모든 객체의 필드에 대하여 `this` 키워드를 붙이는 것이 좋다는 말씀이실까요?

pkeugine replied:
히스타가 생각한 “this 키워드가 꼭 필요한 상황”은 언제인가요? 파라미터와 필드의 이름이 겹칠 때 인가요?
말해주신 것처럼 필수가 아니더라도 this 를 붙였을 때 명확하게 그게 필드인지 아닌지를 드러낼 수 있다고 생각해요!

pkeugine commented on src/main/java/domain/attendance/AttendanceHistory.java:
```diff
@@ -0,0 +1,42 @@
+    public boolean isPastHistory(int day) {
+        return attendanceDateTime.getDay() < day;
+    }
+}
```
파일마다 새로운 line 을 넣어주는게 어떨까요?
현재 PR 에서 변경점을 보면 일부 파일에 아래와 같이 빨간 아이콘이 뜨거든요 :
<img width="78" alt="스크린샷 2025-02-14 오전 2 34 06" src="https://github.com/user-attachments/assets/90b17a4b-ef1f-46b7-a4df-b36cf4c2328f" />

intelliJ 설정에서 개발자가 변경하는 파일마다 자동으로 파일 마지막에 new line 을 넣어주는 기능이 있어요.
이를 사용해보셔도 좋아요!

Hacanna42 replied:
쉽게 빠트릴 수 있는 부분인데 짚어주셔서 감사합니다! 해당 설정 적용했습니다 :)

pkeugine commented on src/main/java/domain/attendance/AttendanceHistories.java:
```diff
@@ -0,0 +1,42 @@
+                .orElseThrow(() -> new IllegalArgumentException("해당 출석 기록이 존재하지 않습니다."));
+    }
+
+    public List<AttendanceHistory> findAllHistoriesOf(Crew crew) {
```
사용되지 않는 메소드가 남아있네요!

Hacanna42 replied:
놓친 리팩토링이 많네요... :(
TDD를 처음 적용하며 개발하려니 시간에 너무 쫒겼었어요..!
다음부터는 꼭 개선된 모습 보이겠습니다. 해당 부분은 빠르게 수정하겠습니다!

pkeugine commented on src/main/java/domain/attendance/AttendanceHistories.java:
```diff
@@ -0,0 +1,42 @@
+
+    public AttendanceHistory findHistoryBy(AttendanceHistory newAttendanceHistory) {
+        return attendanceHistories.stream()
+                .filter(history -> history.hasSameDay(newAttendanceHistory) && history.isSameCrew(newAttendanceHistory))
```
`history.hasSameDay(newAttendanceHistory) && history.isSameCrew(newAttendanceHistory)`이 조건은 결국 두 AttendanceHistory가 같은 기록을 의미하는지를 판단하는 부분인데 하나로 합치는게 어떨까요?
`hasSameDay`와 `isSameCrew` 모두 여기서 밖에 쓰이지 않습니다.

Hacanna42 replied:
네. 확실히 하나로 합치면 가독성이 좋아질 것 같아요.
그런데 두 함수를 완전히 하나로 합치는 것보다, 미래의 확장성을 위해서 각 메서드는 private 메서드로 남기고, 그 두개를 실행하는 래퍼 메서드를 만드는 것에 대해서 어떻게 생각하시나요?

pkeugine replied:
꼭 가독성 때문만은 아닙니다. 같은 기록임을 판단하는 로직을 묶어두면, 그 조건이 바뀔 때 관리하기 편하겠죠.
private 메서드로 두고 유지하는건 히스타가 판단해보면 좋을 것 같아요. :)

pkeugine commented on src/main/java/domain/crew/Crew.java:
```diff
@@ -0,0 +1,40 @@
+        this.name = name;
+    }
+
+    public static Crew of(String name) {
```
정적 팩토리 메소드는 어떨 때 사용하는 거라고 생각하시나요? 👀

Hacanna42 replied:
정적 팩토리 메서드는 기본 생성자보다 장점이 많고, 아직까지 단점을 찾지 못해서 저는 사실 대부분의 상황에서 정적 팩토리 메서드를 사용하는 편입니다.

먼저, 해당 코드에서 정적 팩토리 메서드 네이밍을 틀렸네요.. 인자가 하나인 경우에는 of보다는 from이 적절한데 해당 부분은 수정하도록 하겠습니다.

일단 정적 팩토리 메서드를 사용하면 객체 생성의 책임을 외부가 아닌(그러니까 객체 생성 관련 코드를 외부에서 사용하지 않고) 자기 자신의 클래스에 관련 코드를 둘 수 있기 때문에 코드가 깔끔해진다고 생각합니다.

또, new Crew(String name) 도 괜찮지만 정적 팩토리 메서드를 사용하면 생성자에 이름을 부여할 수 있기 때문에 가독성도 더 좋아집니다. Crew.from(String name);  name 에서부터 Crew를 생성하기 때문에, Crew에게 제일 중요한 필드는 String name; 임을 코드에서 쉽게 짐작할 수 있습니다.

그리고 정적 팩토리 메서드는 유연합니다. 기존 생성자는 무조건 메서드 명을 클래스 명과 일치시켜야 하는데, 정적 팩토리 메서드는 그럴 필요가 없으니 다양한 상황에 대해서 쉽게 여러가지 생성 메서드들을 정의할 수 있습니다.

하지만 사실 현재 상황에서는 생성자에서 유연성을 챙길 필요가 없고, 정적 팩토리 메서드로 인해 가독성이 향상되는 부분도 체감이 크게 되지는 않기 때문에 굳이 정적 팩토리 메서드를 사용할 필요는 없다는 것에 동의하고. 페어와도 이것에 대해 의견을 나눈적이 있었는데요.

결론은 현재 상황에서 정적 팩토리 메서드를 사용하는 것에 장점이 있어서 사용하기 보다는, 정적 팩토리 메서드를 사용하지 않을 이유(단점)을 찾지 못해서 사용한다고 정리할 수 있겠습니다!

pkeugine commented on src/main/java/domain/attendance/AttendanceHistory.java:
```diff
@@ -0,0 +1,42 @@
+        return crew.equals(attendanceHistory.crew);
+    }
+
+    public boolean isSameCrew(Crew crew) {
```
단순히 `equals`를 호출하고 같은 응답을 하는데 `isSameCrew`로 감쌀 필요가 있을까요?

Hacanna42 replied:
`private boolean isSameCrew(AttendanceHistory comparedHistory)` 에 대한 피드백인지
`public boolean isSameCrew(Crew crew)` 에 대한 피드백인지 조금 헷갈리네요!

`public boolean isSameCrew(Crew crew)` 라고 이해하고 답변해보겠습니다. 
만약 저기서 `isSameCrew` 메서드를 사용하지 않는다면, `AttendanceHistory`의 `Crew`가 같은 지 비교하기 위해 먼저 필드 `crew`의 `getter`을 만든 뒤 외부에서 `crew.equals()` 로 비교해야 할 것 같습니다. 그러면 데이터의 응집도가 떨어지기 때문에 설계상 별로일 것 같은데, 굳이 그렇게 해야 할 필요가 있을까요?

그리고, 외부에서 `crew`를 받아와서 `crew.equals(crew)` 를 한다고 가정하더라도, `isSameCrew()` 라는 이름을 부여해서 가독성을 챙기는 것도 좋지 않을까요?

pkeugine replied:
`public boolean isSameCrew(Crew crew)`에 달았던 코멘트였습니다. 🙂
답변하신 내용 좋네요. 해당 메서드가 어디서 사용되는지를 생각해보고 남긴 리뷰였는데, 히스타의 주장에 설득되었습니다 👍

pkeugine commented on src/main/java/controller/dto/AttendanceTypeCountDto.java:
```diff
@@ -0,0 +1,24 @@
+        int lateCount,
+        PenaltyType penaltyType
+) {
+    public static AttendanceTypeCountDto of(Crew crew, AttendanceTypeCount attendanceTypeCount,
```
정적 팩토리 메소드의 이름으로 `of`, `from` 두 가지가 있는데, 둘의 차이가 무엇인가요?

Hacanna42 replied:
정말 많은 사람들이 법칙처럼 따르는 정적 팩토리 메소드 네이밍 컨벤션을 따라해봤습니다.
of는 복수의 값을 통한 객체 생성 메소드의 네이밍이고, from은 단일 값을 통한 객체 생성 메소드의 네이밍입니다.

Hacanna42 replied:
추가로 궁금한 점이 있었는데, `java.time`의 `DayOfWeek` Enum은 정적 팩토리 메서드로
`DayOfWeek.of(int dayOfWeek)` 를 사용하더라고요. 왜 여기서는 네이밍 규칙을 어겼는지 궁금해서 관련 정보를 찾아봤는데도 명쾌한 해답을 얻을 수 없었습니다.

물론 네이밍 컨벤션이 절대적인 규칙은 아니지만, 왜 굳이 많은 사람들이 사용하는 공식 Java API의 정적 팩토리 메서드가 컨벤션을 따르지 않았을까요?

혹시라도 관련해서 알고 계시면 답변을 듣고 싶습니다..! 감사합니다 :)

pkeugine replied:
히스타가 말해주신 컨벤션은 “이펙티브 자바”에서 제시한 하나의 방법입니다. 글로벌 스탠다드가 아니기 때문에 꼭 모든 팀이 적용할 필요는 없습니다.

그렇기에 Java 가 네이밍 규칙을 어겼다고 하기엔 무리가 있습니다. 책이 나오기도 전에 java.time 패키지가 먼저 만들어졌을테니까요.

오히려 해당 책이 나왔음에도 불구하고 java 진영은 왜 섣불리 메서드명을 바꾸지 않을까? 에 대해 고민해보면 재밌을 것 같아요.

pkeugine commented on src/main/java/domain/attendance/AttendanceTypeCount.java:
```diff
@@ -0,0 +1,63 @@
+
+    public static AttendanceTypeCount from(int day, List<AttendanceHistory> attendanceHistories) {
+        Map<AttendanceType, Integer> attendanceTypeCount = new HashMap<>();
+        Map<Integer, AttendanceHistory> historyOfDay = new HashMap<>();
```
`Collectors.toMap` 를 활용하면 Map을 만들고 put 하는 반복문 없이 한 번에 선언이 가능합니다!

Hacanna42 replied:
너무 좋은 팁 감사드립니다! 해당 부분 학습중입니다!
`Stream`에서 `Collectors.toMap`으로 `key`와 `value`를 바로 넣을 수 있고, 
`Function.identity()` 라는 것도 처음 알게 되었네요. 감사합니다 피케이 :)

pkeugine commented on src/main/java/domain/attendance/AttendanceTypeCount.java:
```diff
@@ -0,0 +1,63 @@
+    public int getAbsenceCount() {
+        return attendanceTypeCount.getOrDefault(AttendanceType.ABSENCE, 0);
+    }
+
```
불필요한 줄바꿈이 있습니다!

Hacanna42 replied:
감사합니다. 리팩토링 놓친 부분이 많네요. 해당 부분 바로 수정했습니다!

pkeugine commented on src/main/java/domain/attendance/PenaltyType.java:
```diff
@@ -0,0 +1,25 @@
+        this.absenceCount = absenceCount;
+    }
+
+    // TODO: Enum 상수 선언 순서에 영향받지 않는 더 나은 방법 고려
```
TODO가 남아있네요!
다른 파일들에도 TODO가 좀 있는데, 확인 후 해결해주세요~ 2단계 제출할 때라도요😀

Hacanna42 replied:
네! TODO 모두 삭제하겠습니다.
디테일하게 신경을 못 썼어서 죄송해요 ... 🥺

pkeugine commented on src/main/java/domain/date/AttendanceDate.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    // TODO: 정적 선언이 맞을까 고민
+    public static boolean isRestDay(int day) {
```
정적 선언을 하면 확장성이 떨어집니다. 당장 1월에도 사용할 수 없는 서비스가 되니까요🥲
요일을 통해 판단하는 것으로 변경해보길 바랍니다!

Hacanna42 replied:
프로젝트 설계를 너무 요구사항에만 집중하고, 확장성을 생각 못했던 것 같아요. 당장 1월에 사용할 수 있다는 건 사실 뻔한 예측인데도 말이에요.. 좋은 피드백 감사드립니다!

pkeugine replied:
12월 한정 프로그램이라는 설명을 추가하셨군요!
나중에 여유가 있다면 이 부분도 개선해보면 꽤 재밌을 듯 합니다. :)

pkeugine commented on src/main/java/domain/date/AttendanceDateTime.java:
```diff
@@ -0,0 +1,55 @@
+
+import domain.attendance.AttendanceType;
+
+public class AttendanceDateTime {
```
시간에 대한 기능을 구현할 때 `java.time.LocalDate`, `java.time.LocalDateTime` 등 자바에서 제공해주는 클래스를 활용하면 더 쉽게 해결할 수 있습니다!

Hacanna42 replied:
맞아요.. 사실 이번 미션 첫 PR 올리고 나서 DateTime을 직접 구현한 사람이 드물더라고요..
공식 API에 대한 지식이 조금 부족했었어요..! 꼭 개선하도록 하겠습니다.

pkeugine commented on src/test/java/AttendanceTimeTest.java:
```diff
@@ -0,0 +1,89 @@
+import org.assertj.core.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+public class AttendanceTimeTest {
```
테스트 코드 보충하기!! 📝📝

Hacanna42 replied:
넵. 보충 하겠습니다!

pkeugine reviewed (changes requested):
안녕하세요 히스타!
두 번째 미션 진행하시느라 고생하셨습니다.
이번 미션은 워낙 요구사항이 많아서 시간이 좀 부족하셨던 것 같아요. 

말씀하신 것처럼 TDD 에는 장단점이 있습니다.
그러나 TDD 도 중요하지만, 일단 요구사항을 만족시키고 불필요한 주석 및 TODO 를 포함하지 않도록 해보면 좋을 것 같네요. 다음 리뷰 때 더욱 발전한 모습을 보여주세요.

리뷰 한 번 확인해보시고, 반영 & 리팩토링 하신 뒤 다시 요청 바랍니다.
파이팅입니다 :muscle: :muscle: :muscle:

Hacanna42 committed:
feat: 예외 관련 요구사항 구현

출석 미션은 12월 시범 운영을 위한 프로그램을 작성하는 미션입니다. 하지만 현재 날짜가 2월이기에, 프로그램 첫 시작때 현재 날짜를 지정할 수 있도록 했습니다.

- 현재 날짜를 출력하는 기능 추가
- 이미 출석했는데 다시 출석하는 경우에 대한 예외 처리
- 주말 및 공휴일에 출석하는 경우에 대한 예외 처리
- View 관련 요구사항 구현

Hacanna42 committed:
refactor: 주요 코드 리팩토링

피드백 받은 리팩토링 사항들에 대해 수정했습니다.
- AttendanceHistory의 불필요한 파라미터 제거
- this 키워드 관련 컨벤션 통일
- POSIX EOF 표준 준수
- hasSameDay()와 isSameCrew()의 네이밍 통일 및 병합
- 잘못된 정적 팩토리 메서드 네이밍 수정
- Collectors.toMap() 을 사용하여 불필요한 반복문 제거
- 불필요한 줄바꿈 및 임포트 삭제

Hacanna42 committed:
test: 리팩토링으로 인해 깨진 테스트 수정

Hacanna42 committed:
test: Date 관련 테스트 코드 추가 및 수정

Hacanna42 committed:
test: AttendanceHistoriesTest 추가

Hacanna42 committed:
test: AttendanceHistories, TypeCount 테스트 코드 추가

Hacanna42 committed:
refactor: Controller 가독성 향상 리팩토링

- Function Enum의 이름을 FeatureType으로 변경
- 일부 메서드 분리로 가독성 향상

pkeugine reviewed (approved):
안녕하세요 히스타!
step 1 진행하시느라 고생하셨습니다.
많은 고민과 생각을 거쳐 코드를 작성하시는 점이 인상 깊었어요.
추가 피드백 남겼으니 한 번 확인해주시고,
step 2 진행하면서 반영해보면 좋을듯 합니다.
조만간 다시 뵐게요. 파이팅입니다 🔥 🔥 🔥
